infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationManager.java

```diff
@@ -0,0 +1,300 @@
+            topologyDescriptor.removeStandbySite(siteID);
+        }
+
+        //Start the standbys that are in the new config but not in the current config
+        for (String clusterId : newConfig.getStandbyClusters().keySet()) {
+            if (runtimeToRemoteCluster.get(clusterId) == null) {
+                ClusterDescriptor clusterInfo = newConfig.getStandbyClusters().get(clusterId);
+                topologyDescriptor.addStandbySite(clusterInfo);
+                startLogReplicationRuntime(clusterInfo);
+            }
+        }
+    }
+
+    /**
+     * Query max stream tail for all streams to be replicated.
+     *
+     * @return max tail of all relevant streams.
+     */
+    private long queryStreamTail() {
+        Set<String> streamsToReplicate = logReplicationConfig.getStreamsToReplicate();
+        long maxTail = Address.NON_ADDRESS;
+        for (String s : streamsToReplicate) {
+            UUID currentUUID = CorfuRuntime.getStreamID(s);
+            Map<UUID, Long> tailMap = corfuRuntime.getAddressSpaceView().getAllTails().getStreamTails();
+            Long currentTail = tailMap.get(currentUUID);
+            if (currentTail != null) {
+                maxTail = Math.max(maxTail, currentTail);
+            }
+        }
+        return maxTail;
+    }
+
+    /**
+     * Given a timestamp, calculate how many entries to be sent for all replicated streams.
+     *
+     * @param timestamp
+     */
+    private long queryEntriesToSend(long timestamp) {
+        int totalNumEntries = 0;
+
+        for (CorfuLogReplicationRuntime runtime: runtimeToRemoteCluster.values()) {
+            totalNumEntries += runtime.getNumEntriesToSend(timestamp);
+        }
+
+        return totalNumEntries;
+    }
+
+    /**
+     * Query the current all replication stream log tail and remember the max stream tail.
+     * Query each standby site information according to the ack information to calculate the number of
+     * msgs to be sent out.
+     */
+    public void prepareSiteRoleChange() {
+        prepareSiteRoleChangeStreamTail = queryStreamTail();
+        totalNumEntriesToSend = queryEntriesToSend(prepareSiteRoleChangeStreamTail);
+    }
+
+    /**
+     * Query the all replication stream log tail and calculate the number of messages to be sent.
+     * If the max tail has changed, give 0 percent has done.
+     *
+     * @return Percentage of work has been done, when it return 100, the replication is done.
+     */
+    public int queryReplicationStatus() {
+        long maxTail = queryStreamTail();
+
+        /*
+         * If the tail has been moved, reset the base calculation
+         */
+        if (maxTail > prepareSiteRoleChangeStreamTail) {
+            prepareSiteRoleChange();
+        }
+
+        long currentNumEntriesToSend = queryEntriesToSend(prepareSiteRoleChangeStreamTail);
+        log.debug("maxTail {} totalNumEntriesToSend  {}  currentNumEntriesToSend {}", maxTail, totalNumEntriesToSend, currentNumEntriesToSend);
```

Comment:
if the max stream tail moves, it calls `prepareSiteRoleChange()`, which call `queryStreamTail()` one more time, and will update `totalNumEntriesToSend`. Then it call `queryEntriesToSend()` again, will get a pretty close result as `currentNumEntriesToSend`. So percent calculation will always return a 0.

Reply:
@xiaoqin2012 please address this

Reply:
Adding as a TODO in code for you guys to work on this on a separate PR.